evm/auxiliary/user_account_contract/transaction.py

```diff
@@ -0,0 +1,210 @@
+
+    @vrs.setter
+    def vrs(self, value):
+        for i in value:
+            validate_uint256(i)
+
+        b = b''.join([
+            pad32(int_to_big_endian(i)) for i in value
+        ])
+
+        self.data = self.data[:SIGNATURE_SLICE.start] + b + self.data[SIGNATURE_SLICE.stop:]
+
+    @property
+    def nonce(self):
+        return big_endian_to_int(self.data[NONCE_SLICE])
+
+    @nonce.setter
+    def nonce(self, value):
+        validate_uint256(value)
+        b = pad32(int_to_big_endian(value))
+        self.data = self.data[:NONCE_SLICE.start] + b + self.data[NONCE_SLICE.stop:]
+
+    @property
+    def min_block(self):
+        return big_endian_to_int(self.data[MIN_BLOCK_SLICE])
+
+    @min_block.setter
+    def min_block(self, value):
+        validate_uint256(value)
+        b = pad32(int_to_big_endian(value))
+        self.data = self.data[:MIN_BLOCK_SLICE.start] + b + self.data[MIN_BLOCK_SLICE.stop:]
+
+    @property
+    def max_block(self):
+        return big_endian_to_int(self.data[MAX_BLOCK_SLICE])
+
+    @max_block.setter
+    def max_block(self, value):
+        validate_uint256(value)
+        b = pad32(int_to_big_endian(value))
+        self.data = self.data[:MAX_BLOCK_SLICE.start] + b + self.data[MAX_BLOCK_SLICE.stop:]
+
+    # TODO: rename to gas_price once gas price is removed from ShardingTransaction
+    @property
+    def int_gas_price(self):
+        return big_endian_to_int(self.data[GASPRICE_SLICE])
+
+    @int_gas_price.setter
+    def int_gas_price(self, value):
+        validate_uint256(value)
+        b = pad32(int_to_big_endian(value))
+        self.data = self.data[:GASPRICE_SLICE.start] + b + self.data[GASPRICE_SLICE.stop:]
+
+    @property
+    def value(self):
+        return big_endian_to_int(self.data[VALUE_SLICE])
+
+    @value.setter
+    def value(self, value_):
+        validate_uint256(value_)
+        b = pad32(int_to_big_endian(value_))
+        self.data = self.data[:VALUE_SLICE.start] + b + self.data[VALUE_SLICE.stop:]
+
+    @property
+    def destination(self):
+        destination_padded = self.data[DESTINATION_SLICE]
+        assert destination_padded.startswith(b'\x00' * 12)
+        return destination_padded[12:]
+
+    @destination.setter
+    def destination(self, value):
+        validate_canonical_address(value)
+        b = pad32(value)
+        self.data = self.data[:DESTINATION_SLICE.start] + b + self.data[DESTINATION_SLICE.stop:]
+
+    @property
+    def msg_data(self):
+        return self.data[DATA_SLICE]
+
+    @msg_data.setter
+    def msg_data(self, value):
+        validate_is_bytes(value)
+        self.data = self.data[:DATA_SLICE.start] + value
+
+    #
+    # Signing
+    #
+
+    @property
+    def is_signed(self):
+        return self.vrs != (0, 0, 0)
+
+    def get_message_for_signing(self):
+        return b''.join([
+            self.data[SIGNATURE_SLICE.stop:],
+            self.sig_hash,
+        ])
+
+    def sign(self, private_key):
```

Comment:
Was it intentional to deviate from the existing API for transaction signing which uses `as_signed_transaction` which returns a new object rather than mutating the existing object?

This question is from a broader question about the use of property setters.

- **if** the use of property setters was to embed validation as part of setting the value, then I would suggest changing to use a single `validate()` method which is inline with the other transaction classes.
- **if** the use of property setters was to allow mutation of the object, I would like to understand why since we've trended away from most forms of object mutation.

Reply:
I added the setters to allow mutation. The other transaction objects are technically mutable as well and I wanted to mirror that. But if they aren't mutated in practice I'll be happy to remove them because all these setters look quite ugly. I'll also try to implement the `as_signed_transaction` pattern.